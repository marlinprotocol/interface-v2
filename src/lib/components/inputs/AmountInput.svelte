<script lang="ts">
	export let disabled = false;
	export let value = '';
	export let onChange: (e: Event) => void = () => {};
	export let styleClass = '';
	export let onlyInteger = false;
	export let id = '';
	export let placeholder = '';

	const validateInput = (
		e: KeyboardEvent & {
			currentTarget: EventTarget & HTMLInputElement;
		}
	) => {
		// user can only enter numbers, arrows, backspace, delete and one decimal point if onlyInteger is false
		if (
			(e.key >= '0' && e.key <= '9') ||
			(!onlyInteger && e.key === '.' && !value.includes('.')) ||
			e.key === 'Backspace' ||
			e.key === 'Delete' ||
			e.key === 'ArrowLeft' ||
			e.key === 'ArrowRight' ||
			e.key === 'ArrowUp' ||
			e.key === 'ArrowDown'
		) {
			return true;
		} else {
			e.preventDefault();
		}
	};
</script>

<input
	{disabled}
	id="pond-input-amount-{id}"
	bind:value
	on:input={onChange}
	on:keydown={validateInput}
	on:keyup={validateInput}
	class="hideInputNumberAppearance {styleClass}"
	placeholder={onlyInteger ? '0' : placeholder !== '' ? placeholder : '0.00'}
	autocomplete="off"
/>
