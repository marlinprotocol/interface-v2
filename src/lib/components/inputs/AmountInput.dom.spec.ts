import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import AmountInput from './AmountInput.svelte';

const dispatchKeyDown = (input: HTMLInputElement, key: string, keyCode: string | number) => {
	fireEvent.keyDown(input, { key, keyCode });
};
describe('AmountInput', () => {
	it('renders with correct attributes', () => {
		const { container, getByPlaceholderText } = render(AmountInput, {
			props: {
				disabled: false,
				onlyInteger: false,
				value: '',
				styleClass: 'test-class'
			}
		});

		const input = getByPlaceholderText('0.00');
		expect(input).toBeTruthy();
		expect(input.getAttribute('id')).toBe('pond-input-amount-');
		expect(input.className).toBe('hideInputNumberAppearance test-class');
		expect(container).toMatchSnapshot();
	});

	it('handles onChange event', async () => {
		const mockOnChange = vi.fn();
		const { getByPlaceholderText } = render(AmountInput, {
			props: {
				disabled: false,
				onlyInteger: false,
				value: '',
				onChange: mockOnChange
			}
		});

		const input = getByPlaceholderText('0.00') as HTMLInputElement;
		await fireEvent.input(input, { target: { value: '123' } });
		expect(mockOnChange).toHaveBeenCalled();
		expect(input.value).toBe('123');
	});

	it('allows numeric input without decimal point when onlyInteger is true', async () => {
		const { getByPlaceholderText } = render(AmountInput, { onlyInteger: true });
		const input = getByPlaceholderText('0') as HTMLInputElement;
		dispatchKeyDown(input, '5', 53); // Simulate pressing the '5' key
		expect(input.value).toBe(''); // The value should still be empty because pressing the key itself won't update value
	});

	it('prevents non-numeric input when onlyInteger is true', async () => {
		const { getByPlaceholderText } = render(AmountInput, { onlyInteger: true });
		const input = getByPlaceholderText('0') as HTMLInputElement;
		dispatchKeyDown(input, '.', 190); // Simulate pressing the '.' key
		expect(input.value).toBe(''); // Value should not change since '.' is not allowed for onlyInteger = true
	});

	it('allows numeric input and a single decimal point when onlyInteger is false', async () => {
		const { getByPlaceholderText } = render(AmountInput, { onlyInteger: false });
		const input = getByPlaceholderText('0.00') as HTMLInputElement;
		dispatchKeyDown(input, '3', 51); // Simulate pressing the '3' key
		dispatchKeyDown(input, '.', 190); // Simulate pressing the '.' key
		expect(input.value).toBe(''); // The value should still be empty because pressing the key itself won't update value
	});
});
