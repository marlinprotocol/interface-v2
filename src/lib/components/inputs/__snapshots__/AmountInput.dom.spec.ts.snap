// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AmountInput > renders with correct attributes 1`] = `
<div>
  <input
    autocomplete="off"
    class="hideInputNumberAppearance test-class"
    id="pond-input-amount-"
    placeholder="0.00"
  />
</div>
`;
